src/vm/eventpipe.cpp

```diff
@@ -0,0 +1,234 @@
+#include "eventpipejsonfile.h"
+#include "sampleprofiler.h"
+
+CrstStatic EventPipe::s_initCrst;
+bool EventPipe::s_tracingInitialized = false;
+bool EventPipe::s_tracingEnabled = false;
+EventPipeJsonFile* EventPipe::s_pJsonFile = NULL;
+
+void EventPipe::Initialize()
+{
+    STANDARD_VM_CONTRACT;
+
+    s_tracingInitialized = s_initCrst.InitNoThrow(
+        CrstEventPipe,
+        (CrstFlags)(CRST_TAKEN_DURING_SHUTDOWN));
+}
+
+void EventPipe::EnableOnStartup()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    // Test COMPLUS variable to enable tracing at start-up.
+    if(CLRConfig::GetConfigValue(CLRConfig::INTERNAL_PerformanceTracing) != 0)
+    {
+        Enable();
+    }
+}
+
+void EventPipe::Shutdown()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    Disable();
+}
+
+void EventPipe::Enable()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    if(!s_tracingInitialized)
+    {
+        return;
+    }
+
+    // Take the lock and enable tracing.
+    CrstHolder _crst(&s_initCrst);
+    s_tracingEnabled = true;
+    if(CLRConfig::GetConfigValue(CLRConfig::INTERNAL_PerformanceTracing) == 2)
+    {
+        // File placed in current working directory.
+        SString outputFilePath;
+        outputFilePath.Printf("Process-%d.PerfView.json", GetCurrentProcessId());
+        s_pJsonFile = new EventPipeJsonFile(outputFilePath);
+    }
+
+    SampleProfiler::Enable();
+}
+
+void EventPipe::Disable()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    CrstHolder _crst(&s_initCrst);
+    s_tracingEnabled = false;
+    SampleProfiler::Disable();
+
+    if(s_pJsonFile != NULL)
+    {
+        delete(s_pJsonFile);
+        s_pJsonFile = NULL;
+    }
+}
+
+bool EventPipe::EventEnabled(GUID& providerID, INT64 keyword)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    // TODO: Implement filtering.
+    return false;
+}
+
+void EventPipe::WriteEvent(GUID& providerID, INT64 eventID, BYTE *pData, size_t length, bool sampleStack)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    StackContents stackContents;
+    bool stackWalkSucceeded;
+
+    if(sampleStack)
+    {
+        stackWalkSucceeded = WalkManagedStackForCurrentThread(stackContents);
+    }
+
+    // TODO: Write the event.
+}
+
+void EventPipe::WriteSampleProfileEvent(Thread *pThread, StackContents &stackContents)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+        PRECONDITION(pThread != NULL);
+    }
+    CONTRACTL_END;
+
+    EX_TRY
+    {
+        if(s_pJsonFile != NULL)
+        {
+            static CommonEventFields eventFields;
```

Comment:
I was going for fewer allocations, but this actually would have turned into a bug as soon as multiple callers exist.  I've fixed this.